### DISCUSSION TITLE
How to dispose of AvaloniaUI Application

### DISCUSSION THREAD

**Discussion by matthewcrews** [fsprojects/Avalonia.FuncUI#331] [Q&A] [answered]
I am using AvaloniuaFuncUI to create a add-in for Excel. I can open a window and close it using the "X" button in the top right of the window. When I go to open it again I get the following error:

> Can not have multiple active ClassicDesktopStyleApplicationLifetime instances and the previously created one was not disposed

What is the correct way to handle this situation? I have searched the API for a way to dispose of the application and have not found one. I have also tried to open the main window again instead of launching the application again but have not found a way to do that either.

Is there any documentation or guides for a scenario where you want the application to stay open and just hide/show the main window?

![image](https://github.com/fsprojects/Avalonia.FuncUI/assets/1593215/2ace9a97-9f1e-48b1-af9f-5ca5dd6054a9)

**Comment by JordanMarr** [comment, marked as answer]
Hi Matthew!

In the world of add-ins, the AppDomain belongs to the parent app (Excel), so you have to be careful not to call your app init code more than once. 

First of all, you have some control over the conditions that shutdown your app via the `ShutdownMode` property:

```F#
    override this.OnFrameworkInitializationCompleted() =
        match this.ApplicationLifetime with
        | :? IClassicDesktopStyleApplicationLifetime as desktop ->         
            let view = MainView()
            desktop.MainWindow <- view
            desktop.ShutdownMode <- ShutdownMode.OnMainWindowClose
```

Values are:
* `ShutdownMode.OnMainWindowClose`
* `ShutdownMode.OnExplicitShutdown`
* `ShutdownMode.OnLastWindowClose`


I've had to do this in the past with Revit add-ins:
```C#
if (Application.Current == null)
{
    var app = new Application();
    app.ShutdownMode = ShutdownMode.OnExplicitShutdown;
}
```

I know that's not a direct answer, but I hope that it's enough of a hint to put you on the right track!
Please share the final recipe if you get it working for future reference.

**Reply by matthewcrews** [reply]
Thank you. I already have something like this.

```
type App() =
    inherit Application()

    override this.Initialize() =
        this.Styles.Add (FluentTheme())
        this.RequestedThemeVariant <- Styling.ThemeVariant.Dark

    override this.OnFrameworkInitializationCompleted() =
        match this.ApplicationLifetime with
        | :? IClassicDesktopStyleApplicationLifetime as desktopLifetime ->
            let mainWindow = MainWindow()
            desktopLifetime.MainWindow <- mainWindow
            desktopLifetime.ShutdownMode <- Controls.ShutdownMode.OnMainWindowClose
        | _ -> ()

let appBuilder =
    AppBuilder
        .Configure<App>()
        .UsePlatformDetect()
        .UseSkia()
        .SetupWithoutStarting()

let showEditor () =
    let d (_: Application) (_: string[]) = ()
    let myDelegate = AppBuilder.AppMainDelegate d
    appBuilder.Start(myDelegate, [||])
    ()
```

The `showEditor` function is the function called from Excel to bring up the AvaloniaUI window.

I think I have two main issues: 
1. I don't have a way to detect whether the AvaloniaUI app is already running
2. I don't know how to re-open the Main Window when it has been closed.

**Reply by JordanMarr** [reply]
Can you check the `Application.Current` static property for null?

**Reply by JordanMarr** [reply]
You might also try `ShutdownMode.OnExplicitShutdown` mode so that it doesn't shut down automatically when the window closes.

**Reply by matthewcrews** [reply]
With the following code I am now getting an error saying that I'm calling the `Setup` more than once.

```
type App() =
    inherit Application()

    override this.Initialize() =
        this.Styles.Add (FluentTheme())
        this.RequestedThemeVariant <- Styling.ThemeVariant.Dark

    override this.OnFrameworkInitializationCompleted() =
        match this.ApplicationLifetime with
        | :? IClassicDesktopStyleApplicationLifetime as desktopLifetime ->
            let mainWindow = MainWindow()
            desktopLifetime.MainWindow <- mainWindow
            desktopLifetime.ShutdownMode <- Controls.ShutdownMode.OnExplicitShutdown
        | _ -> ()

let appBuilder =
    AppBuilder
        .Configure<App>()
        .UsePlatformDetect()
        .UseSkia()
        .SetupWithoutStarting()

let showEditor () =
    match Application.Current with
    | null ->
        let d (_: Application) (_: string[]) = ()
        let myDelegate = AppBuilder.AppMainDelegate d
        appBuilder.Start(myDelegate, [||])

    | _ ->
        ()
    ()
```
![image](https://github.com/fsprojects/Avalonia.FuncUI/assets/1593215/782c1d18-f220-4c67-9c68-b6808ee17dc7)

**Reply by JordanMarr** [reply]
With WPF, I seem to recall that there were different ways that `Application` could be started.
The default templates would have an `App.xaml` file with its `Build Action` set in a certain way that caused it to automatically load. In Avalonia, it is set to `Avalonia XAML`. I'm guessing it probably uses reflection to see if you have an `Application` and then auto starts it.
I believe I removed that file altogether and just called my own class that manually created a new `Application` if `Application.Current == null`. 

That might be a bit of a rabbit hole though. At the end of the day, a new `Application` was only created once within the lifetime of the host application (Revit, Excel, etc).
